File: ml_test_vectors/stateless/sum/implementation.py
from functools import reduce
from typing import Optional

import numpy as np

from ml_test_vectors.types import Scalar, Tensor


def forward(inp: Tensor) -> Scalar:
    return (
        inp
        if isinstance(inp, float)
        else reduce(lambda subtotal, m: subtotal + forward(m), inp, 0)
    )


def backward(inp: Tensor, _output: Optional[Tensor]) -> Tensor:
    return np.ones_like(inp)

File: ml_test_vectors/stateless/sum/implementation.ts
import type {
	Scalar,
	StatelessDerivativeFunction,
	StatelessFunction,
	Tensor,
} from "../../types";
import { isScalar, onesLike } from "../../utils";

export const forward: StatelessFunction<Tensor, Scalar> = (input) => {
	if (isScalar(input)) {
		return input;
	}
	return input.reduce<number>((subtotal, n) => subtotal + forward(n), 0);
};

export const backward: StatelessDerivativeFunction = (input) => onesLike(input);
